pkg/yurthub/otaupdate/ota.go

```diff
@@ -0,0 +1,148 @@
+		podList, err := clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{
+			FieldSelector: "spec.nodeName=" + nodeName,
+		})
+		if err != nil {
+			klog.Errorf("Get pod list failed, %v", err)
+			returnErr(fmt.Errorf("Get pods list failed"), w, http.StatusInternalServerError)
+			return
+		}
+		klog.V(5).Infof("Got pod list: %v", podList)
+
+		// Successfully get pod list
+		w.Header().Set("content-type", "text/json")
+		data, err := json.Marshal(podList)
+		if err != nil {
+			klog.Errorf("Marshal pod list failed: %v", err.Error())
+			returnErr(fmt.Errorf("Get pod list failed: data transfer to json format failed."), w, http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		n, err := w.Write(data)
+		if err != nil || n != len(data) {
+			klog.Errorf("Write resp for request, expect %d bytes but write %d bytes with error, %v", len(data), n, err)
+		}
+	})
+}
+
+// UpdatePod update a specifc pod(namespace/podname) to the latest version
+func UpdatePod(clientset kubernetes.Interface, nodeName string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		params := mux.Vars(r)
+		namespace := params["ns"]
+		podName := params["podname"]
+
+		err, ok := applyUpdate(clientset, namespace, podName, nodeName)
+		if err != nil {
+			returnErr(fmt.Errorf("Apply update failed"), w, http.StatusInternalServerError)
+			return
+		}
+		if !ok {
+			returnErr(fmt.Errorf("Pod is not-updatable"), w, http.StatusForbidden)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	})
+}
+
+// applyUpdate execute pod update process by deleting pod under OnDelete update strategy
+func applyUpdate(clientset kubernetes.Interface, namespace, podName, nodeName string) (error, bool) {
+	pod, err := clientset.CoreV1().Pods(namespace).Get(context.TODO(), podName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Get pod %v/%v failed, %v", namespace, podName, err)
+		return err, false
+	}
+
+	// Pod will not be updated when it's being deleted
+	if pod.DeletionTimestamp != nil {
+		klog.Infof("Pod %v/%v is deleting, can not be updated", namespace, podName)
+		return nil, false
+	}
+
+	// Pod will not be updated when it's not running on the current node
+	if pod.Spec.NodeName != nodeName {
+		klog.Infof("Pod: %v/%v is running on %v, can not be updated", namespace, podName, pod.Spec.NodeName)
+		return nil, false
+	}
+
+	// Pod will not be updated without pod condition PodNeedUpgrade=true
+	if !IsPodUpdatable(pod) {
+		klog.Infof("Pod: %v/%v is not updatable", namespace, podName)
+		return nil, false
+	}
+
+	klog.V(5).Infof("Pod: %v/%v is updatable", namespace, podName)
+	err = clientset.CoreV1().Pods(namespace).Delete(context.TODO(), podName, metav1.DeleteOptions{})
+	if err != nil {
+		klog.Errorf("Update pod %v/%v failed when delete, %v", namespace, podName, err)
+		return err, false
+	}
+
+	klog.Infof("Update pod: %v/%v success", namespace, podName)
+	return nil, true
```

Comment:
I think do we need to wait for the new version pod to be created and then return?